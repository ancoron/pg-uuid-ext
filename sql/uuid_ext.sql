SET timezone TO 'Zulu';
\x
SELECT extract(epoch from ts), to_char(ts, 'YYYY-MM-DD HH24:MI:SS.US')
FROM uuid_v1_timestamp('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3') AS ts;

SELECT extract(epoch from ts), to_char(ts, 'YYYY-MM-DD HH24:MI:SS.US')
FROM uuid_v1_timestamp('4938f30e-8449-11e9-ae2b-e03f49467033') AS ts;

SELECT
    coalesce(to_char(uuid_v1_timestamp('87c771ce-bc95-3114-ae59-c0e26acf8e81'), 'YYYY-MM-DD HH24:MI:SS.US'), '<null>') AS ver_3,
    coalesce(to_char(uuid_v1_timestamp('22859369-3a4f-49ef-8264-1aaf0a953299'), 'YYYY-MM-DD HH24:MI:SS.US'), '<null>') AS ver_4,
    coalesce(to_char(uuid_v1_timestamp('c9aec822-6992-5c93-b34a-33cc0e952b5e'), 'YYYY-MM-DD HH24:MI:SS.US'), '<null>') AS ver_5,
    coalesce(to_char(uuid_v1_timestamp('00000000-0000-0000-0000-000000000000'), 'YYYY-MM-DD HH24:MI:SS.US'), '<null>') AS nil
;

SELECT
    uuid_version('00000000-0000-0000-0000-000000000000') AS ver_0,
    uuid_version('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3') AS ver_1,
    uuid_version('87c771ce-bc95-3114-ae59-c0e26acf8e81') AS ver_3,
    uuid_version('314dffb7-03ce-442c-b1dd-c1a9b3919282') AS ver_4,
    uuid_version('c9aec822-6992-5c93-b34a-33cc0e952b5e') AS ver_5
;

SELECT
    uuid_variant('00000000-0000-0000-0000-000000000000') AS var_0,
    uuid_variant('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3') AS var_1,
    uuid_variant('{000C1090-0000-0000-C000-000000000046}') AS var_2,
    uuid_variant('C9AEC82269925C93E34A33CC0E952B5E') AS var_3
;

SELECT
    coalesce(uuid_v1_node('b647e96b-862d-11e9-ae2b-db6f0f573554'), '<null>') AS ver_1,
    coalesce(uuid_v1_node('87c771ce-bc95-3114-ae59-c0e26acf8e81'), '<null>') AS ver_3,
    coalesce(uuid_v1_node('22859369-3a4f-49ef-8264-1aaf0a953299'), '<null>') AS ver_4,
    coalesce(uuid_v1_node('c9aec822-6992-5c93-b34a-33cc0e952b5e'), '<null>') AS ver_5,
    coalesce(uuid_v1_node('00000000-0000-0000-0000-000000000000'), '<null>') AS nil
;

SELECT
    uuid_timestamp_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS eq,
    uuid_timestamp_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded3-8dbb-11e9-ae2b-db6f0f573554') AS lt,
    uuid_timestamp_cmp('8385ded3-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS gt,
    uuid_timestamp_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2c-db6f0f573554') AS lt_clock,
    uuid_timestamp_cmp('8385ded2-8dbb-11e9-ae2c-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS gt_clock,
    uuid_timestamp_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573555') AS lt_node,
    uuid_timestamp_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573555', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS gt_node
;

CREATE TABLE uuid_test (id uuid);
CREATE UNIQUE INDEX uidx_uuid_test_id ON uuid_test (id uuid_timestamp_ops);

INSERT INTO uuid_test (id) VALUES
('1004cd50-4241-11e9-b3ab-db6f0f573554'), -- 2019-03-09 07:58:02.056840
('05602550-8a8c-11e9-b3ab-db6f0f573554'), -- 2019-06-09 07:56:00.175240
('8385ded2-8dbb-11e9-ae2b-db6f0f573554'), -- 2019-06-13 09:13:31.650017
('ffc449f0-8c2f-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.391640
('ffc449f0-8c2f-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.391640
('ffc449f0-8c2f-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:19.391640
('ffc449f0-8c2f-11e9-8f34-e03f49c7763b'), -- 2019-06-11 10:02:19.391640
('ffced5f0-8c2f-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.460760
('ffd961f0-8c2f-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.529880
('ffe3edf0-8c2f-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:19.599000
('ffee79f0-8c2f-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.668120
('fff905f0-8c2f-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.737240
('000391f0-8c30-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.806360
('000e1df0-8c30-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:19.875480
('0018a9f0-8c30-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.944600
('002335f0-8c30-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:20.013720
('002dc1f0-8c30-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:20.082840
('00384df0-8c30-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:20.151960
('0042d9f0-8c30-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:20.221080
('004d65f0-8c30-11e9-9bb8-e03f4977f7b7')  -- 2019-06-11 10:02:20.290200
;

ANALYZE uuid_test;

SELECT
    count(*) FILTER (WHERE id <* '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_lt,
    count(*) FILTER (WHERE id <=* '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_le,
    count(*) FILTER (WHERE id >* '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_gt,
    count(*) FILTER (WHERE id >=* '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_ge
FROM uuid_test;

SELECT
    count(*) FILTER (WHERE id <~ '2019-06-11 10:02:20.013720') AS count_lt,
    count(*) FILTER (WHERE id <=~ '2019-06-11 10:02:20.013720') AS count_le,
    count(*) FILTER (WHERE id >~ '2019-06-11 10:02:20.013720') AS count_gt,
    count(*) FILTER (WHERE id >=~ '2019-06-11 10:02:20.013720') AS count_ge
FROM uuid_test;

SET enable_seqscan TO off;
\x
SET timezone TO 'Asia/Tokyo';
EXPLAIN (ANALYZE, TIMING OFF, SUMMARY OFF, COSTS OFF)
SELECT count(*) FROM uuid_test WHERE id <~ '2019-06-11 10:02:19Z';

EXPLAIN (ANALYZE, TIMING OFF, SUMMARY OFF, COSTS OFF)
SELECT * FROM uuid_test WHERE id = '000e1df0-8c30-11e9-9bb8-e03f4977f7b7';
