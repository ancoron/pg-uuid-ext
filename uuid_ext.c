/*
 * Copyright 2019 Ancoron Luciferis
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <sys/time.h>
#include <sys/types.h>
#include <unistd.h>
#include <math.h>

#include "postgres.h"
#include "fmgr.h"
#include "port.h"
#include "datatype/timestamp.h"
#include "lib/stringinfo.h"
#include "utils/builtins.h"
#include "utils/uuid.h"
#include "utils/timestamp.h"

#define UUID_TIME_OFFSET            INT64CONST(122192928000000000)

#define UUID_VARIANT_NCS            (0x00)
#define UUID_VARIANT_RFC4122        (0x01)
#define UUID_VARIANT_GUID           (0x02)
#define UUID_VARIANT_FUTURE         (0x03)

#define UUID_VERSION(uuid)              ((uuid->data[6] >> 4) & 0x0F)
#define UUID_VARIANT_IS_NCS(uuid)       (((uuid->data[8]) & 0x80) == 0x00)
#define UUID_VARIANT_IS_RFC4122(uuid)   (((uuid->data[8]) & 0xC0) == 0x80)
#define UUID_VARIANT_IS_GUID(uuid)      (((uuid->data[8]) & 0xE0) == 0xC0)
#define UUID_VARIANT_IS_FUTURE(uuid)    (((uuid->data[8]) & 0xE0) == 0xE0)

PG_MODULE_MAGIC;

static uint8 uuid_version_internal(pg_uuid_t *uuid);
static uint8 uuid_variant_internal(pg_uuid_t *uuid);
static bool uuid_is_rfc_v1_internal(pg_uuid_t *uuid);

PG_FUNCTION_INFO_V1(uuid_v1_timestamp);
PG_FUNCTION_INFO_V1(uuid_version);
PG_FUNCTION_INFO_V1(uuid_variant);
PG_FUNCTION_INFO_V1(uuid_v1_node);

static uint8
uuid_version_internal(pg_uuid_t *uuid)
{
    return UUID_VERSION(uuid);
}

static uint8
uuid_variant_internal(pg_uuid_t *uuid)
{
    if (UUID_VARIANT_IS_RFC4122(uuid))
        return UUID_VARIANT_RFC4122;

    if (UUID_VARIANT_IS_GUID(uuid))
        return UUID_VARIANT_GUID;

    if (UUID_VARIANT_IS_NCS(uuid))
        return UUID_VARIANT_NCS;

    return UUID_VARIANT_FUTURE;
}

static bool
uuid_is_rfc_v1_internal(pg_uuid_t *uuid)
{
    return 1 == UUID_VERSION(uuid) && UUID_VARIANT_IS_RFC4122(uuid);
}

/*
 * uuid_to_timestamp
 *	extract the timestamp of a version 1 UUID
 *
 */
Datum
uuid_v1_timestamp(PG_FUNCTION_ARGS)
{
    TimestampTz         timestamp = 0L;
	pg_uuid_t           *uuid = PG_GETARG_UUID_P(0);

    if (PG_ARGISNULL(0))
        PG_RETURN_NULL();

    /* version and variant check */
    if (!uuid_is_rfc_v1_internal(uuid))
        PG_RETURN_NULL();

    /* extract shuffled 60 bits to get the UUID timestamp */
    timestamp |= ( ((int64) uuid->data[6] & 0x0F) << 56 );
    timestamp |= ( ((int64) uuid->data[7]) << 48 );
    timestamp |= ( ((int64) uuid->data[4]) << 40 );
    timestamp |= ( ((int64) uuid->data[5]) << 32 );
    timestamp |= ( ((int64) uuid->data[0]) << 24 );
    timestamp |= ( ((int64) uuid->data[1]) << 16 );
    timestamp |= ( ((int64) uuid->data[2]) << 8 );
    timestamp |= ( (int64) uuid->data[3] );

    /* from 100 ns precision to PostgreSQL epoch */
    timestamp -= UUID_TIME_OFFSET;
    timestamp /= 10;
    timestamp -= ((POSTGRES_EPOCH_JDATE - UNIX_EPOCH_JDATE) * SECS_PER_DAY * USECS_PER_SEC);

    /* Recheck in case roundoff produces something just out of range */
    if (!IS_VALID_TIMESTAMP(timestamp))
        ereport(ERROR,
                (errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
                 errmsg("timestamp out of range: \"%ld\"", timestamp)));

	PG_RETURN_TIMESTAMP(timestamp);
}

/*
 * uuid_version
 *	extract the version of a UUID
 *
 */
Datum
uuid_version(PG_FUNCTION_ARGS)
{
	pg_uuid_t *uuid = PG_GETARG_UUID_P(0);

    if (PG_ARGISNULL(0))
        PG_RETURN_NULL();

    PG_RETURN_TIMESTAMP(uuid_version_internal(uuid));
}

/*
 * uuid_version
 *	extract the version of a UUID
 *
 */
Datum
uuid_variant(PG_FUNCTION_ARGS)
{
	pg_uuid_t *uuid = PG_GETARG_UUID_P(0);

    if (PG_ARGISNULL(0))
        PG_RETURN_NULL();

    PG_RETURN_TIMESTAMP(uuid_variant_internal(uuid));
}

/*
 * uuid_v1_node
 *	extract the node of a version 1 UUID (returns NULL otherwise)
 *
 */
Datum
uuid_v1_node(PG_FUNCTION_ARGS)
{
	pg_uuid_t        *uuid = PG_GETARG_UUID_P(0);
	static const char hex_chars[] = "0123456789abcdef";
	char*             node = malloc(13);
	int               i;
	int               j = 0;

    if (PG_ARGISNULL(0))
        PG_RETURN_NULL();

    /* version and variant check */
    if (!uuid_is_rfc_v1_internal(uuid))
        PG_RETURN_NULL();

	for (i = 10; i < UUID_LEN; i++)
	{
		int			hi;
		int			lo;

		hi = uuid->data[i] >> 4;
		lo = uuid->data[i] & 0x0F;

		node[j++] = hex_chars[hi];
		node[j++] = hex_chars[lo];
	}

    node[j++] = '\0';

	PG_RETURN_TEXT_P(cstring_to_text(node));
}
