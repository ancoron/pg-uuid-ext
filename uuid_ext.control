# Extended UUID functions
comment = 'UUID functions'
default_version = '0.3'
module_pathname = '$libdir/uuid_ext'
relocatable = true
