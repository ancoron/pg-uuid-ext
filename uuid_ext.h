/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *      uuid_ext.h
 */

#ifndef UUID_EXT_H
#define UUID_EXT_H

/*
 * The time offset between the UUID timestamp and the PostgreSQL epoch in
 * microsecond precision.
 *
 * This constant is the result of the following expression:
 * `122192928000000000 / 10 + ((POSTGRES_EPOCH_JDATE - UNIX_EPOCH_JDATE) * SECS_PER_DAY * USECS_PER_SEC)`
 */
#define PG_UUID_OFFSET              INT64CONST(13165977600000000)

#define UUID_VARIANT_NCS            (0x00)
#define UUID_VARIANT_RFC4122        (0x01)
#define UUID_VARIANT_GUID           (0x02)
#define UUID_VARIANT_FUTURE         (0x03)

#define UUID_VERSION(uuid)              ((uuid->data[6] >> 4) & 0x0F)
#define UUID_VARIANT_IS_NCS(uuid)       (((uuid->data[8]) & 0x80) == 0x00)
#define UUID_VARIANT_IS_RFC4122(uuid)   (((uuid->data[8]) & 0xC0) == 0x80)
#define UUID_VARIANT_IS_GUID(uuid)      (((uuid->data[8]) & 0xE0) == 0xC0)
#define UUID_VARIANT_IS_FUTURE(uuid)    (((uuid->data[8]) & 0xE0) == 0xE0)

extern uint8 uuid_get_version(const pg_uuid_t *uuid);
extern uint8 uuid_get_variant(const pg_uuid_t *uuid);
extern bool uuid_is_rfc_v1(const pg_uuid_t *uuid);

#endif /* UUID_EXT_H */

