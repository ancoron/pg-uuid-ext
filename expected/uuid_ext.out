SET timezone TO 'Zulu';
\x
SELECT extract(epoch from ts), to_char(ts, 'YYYY-MM-DD HH24:MI:SS.US')
FROM uuid_v1_timestamp('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3') AS ts;
-[ RECORD 1 ]-------------------------
date_part | 1519600168.09884
to_char   | 2018-02-25 23:09:28.098840

SELECT extract(epoch from ts), to_char(ts, 'YYYY-MM-DD HH24:MI:SS.US')
FROM uuid_v1_timestamp('4938f30e-8449-11e9-ae2b-e03f49467033') AS ts;
-[ RECORD 1 ]-------------------------
date_part | 1559378590.74588
to_char   | 2019-06-01 08:43:10.745883

SELECT
    coalesce(to_char(uuid_v1_timestamp('87c771ce-bc95-3114-ae59-c0e26acf8e81'), 'YYYY-MM-DD HH24:MI:SS.US'), '<null>') AS ver_3,
    coalesce(to_char(uuid_v1_timestamp('22859369-3a4f-49ef-8264-1aaf0a953299'), 'YYYY-MM-DD HH24:MI:SS.US'), '<null>') AS ver_4,
    coalesce(to_char(uuid_v1_timestamp('c9aec822-6992-5c93-b34a-33cc0e952b5e'), 'YYYY-MM-DD HH24:MI:SS.US'), '<null>') AS ver_5,
    coalesce(to_char(uuid_v1_timestamp('00000000-0000-0000-0000-000000000000'), 'YYYY-MM-DD HH24:MI:SS.US'), '<null>') AS nil
;
-[ RECORD 1 ]-
ver_3 | <null>
ver_4 | <null>
ver_5 | <null>
nil   | <null>

SELECT
    uuid_version('00000000-0000-0000-0000-000000000000') AS ver_0,
    uuid_version('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3') AS ver_1,
    uuid_version('87c771ce-bc95-3114-ae59-c0e26acf8e81') AS ver_3,
    uuid_version('314dffb7-03ce-442c-b1dd-c1a9b3919282') AS ver_4,
    uuid_version('c9aec822-6992-5c93-b34a-33cc0e952b5e') AS ver_5
;
-[ RECORD 1 ]
ver_0 | 0
ver_1 | 1
ver_3 | 3
ver_4 | 4
ver_5 | 5

SELECT
    uuid_variant('00000000-0000-0000-0000-000000000000') AS var_0,
    uuid_variant('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3') AS var_1,
    uuid_variant('{000C1090-0000-0000-C000-000000000046}') AS var_2,
    uuid_variant('C9AEC82269925C93E34A33CC0E952B5E') AS var_3
;
-[ RECORD 1 ]
var_0 | 0
var_1 | 1
var_2 | 2
var_3 | 3

SELECT
    coalesce(uuid_v1_node('b647e96b-862d-11e9-ae2b-db6f0f573554'), '<null>') AS ver_1,
    coalesce(uuid_v1_node('87c771ce-bc95-3114-ae59-c0e26acf8e81'), '<null>') AS ver_3,
    coalesce(uuid_v1_node('22859369-3a4f-49ef-8264-1aaf0a953299'), '<null>') AS ver_4,
    coalesce(uuid_v1_node('c9aec822-6992-5c93-b34a-33cc0e952b5e'), '<null>') AS ver_5,
    coalesce(uuid_v1_node('00000000-0000-0000-0000-000000000000'), '<null>') AS nil
;
-[ RECORD 1 ]-------
ver_1 | db6f0f573554
ver_3 | <null>
ver_4 | <null>
ver_5 | <null>
nil   | <null>

SELECT
    uuid_timestamp_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS eq,
    uuid_timestamp_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded3-8dbb-11e9-ae2b-db6f0f573554') AS lt,
    uuid_timestamp_cmp('8385ded3-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS gt,
    uuid_timestamp_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2c-db6f0f573554') AS lt_clock,
    uuid_timestamp_cmp('8385ded2-8dbb-11e9-ae2c-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS gt_clock,
    uuid_timestamp_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573555') AS lt_node,
    uuid_timestamp_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573555', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS gt_node
;
-[ RECORD 1 ]
eq       | 0
lt       | -1
gt       | 1
lt_clock | -1
gt_clock | 1
lt_node  | -1
gt_node  | 1

CREATE TABLE uuid_test (id uuid);
CREATE UNIQUE INDEX uidx_uuid_test_id ON uuid_test (id uuid_timestamp_ops);
INSERT INTO uuid_test (id) VALUES
('1004cd50-4241-11e9-b3ab-db6f0f573554'), -- 2019-03-09 07:58:02.056840
('05602550-8a8c-11e9-b3ab-db6f0f573554'), -- 2019-06-09 07:56:00.175240
('8385ded2-8dbb-11e9-ae2b-db6f0f573554'), -- 2019-06-13 09:13:31.650017
('ffc449f0-8c2f-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.391640
('ffc449f0-8c2f-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.391640
('ffc449f0-8c2f-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:19.391640
('ffc449f0-8c2f-11e9-8f34-e03f49c7763b'), -- 2019-06-11 10:02:19.391640
('ffced5f0-8c2f-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.460760
('ffd961f0-8c2f-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.529880
('ffe3edf0-8c2f-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:19.599000
('ffee79f0-8c2f-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.668120
('fff905f0-8c2f-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.737240
('000391f0-8c30-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.806360
('000e1df0-8c30-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:19.875480
('0018a9f0-8c30-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.944600
('002335f0-8c30-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:20.013720
('002dc1f0-8c30-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:20.082840
('00384df0-8c30-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:20.151960
('0042d9f0-8c30-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:20.221080
('004d65f0-8c30-11e9-9bb8-e03f4977f7b7')  -- 2019-06-11 10:02:20.290200
;
ANALYZE uuid_test;
SELECT
    count(*) FILTER (WHERE id <* '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_lt,
    count(*) FILTER (WHERE id <=* '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_le,
    count(*) FILTER (WHERE id >* '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_gt,
    count(*) FILTER (WHERE id >=* '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_ge
FROM uuid_test;
-[ RECORD 1 ]
count_lt | 14
count_le | 15
count_gt | 5
count_ge | 6

SELECT
    count(*) FILTER (WHERE id <~ '2019-06-11 10:02:20.013720') AS count_lt,
    count(*) FILTER (WHERE id <=~ '2019-06-11 10:02:20.013720') AS count_le,
    count(*) FILTER (WHERE id >~ '2019-06-11 10:02:20.013720') AS count_gt,
    count(*) FILTER (WHERE id >=~ '2019-06-11 10:02:20.013720') AS count_ge
FROM uuid_test;
-[ RECORD 1 ]
count_lt | 14
count_le | 15
count_gt | 5
count_ge | 6

SET enable_seqscan TO off;
\x
SET timezone TO 'Asia/Tokyo';
EXPLAIN (ANALYZE, TIMING OFF, SUMMARY OFF, COSTS OFF)
SELECT count(*) FROM uuid_test WHERE id <~ '2019-06-11 10:02:19Z';
                                      QUERY PLAN                                      
--------------------------------------------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   ->  Index Only Scan using uidx_uuid_test_id on uuid_test (actual rows=2 loops=1)
         Index Cond: (id <~ 'Tue Jun 11 19:02:19 2019 JST'::timestamp with time zone)
         Heap Fetches: 2
(4 rows)

EXPLAIN (ANALYZE, TIMING OFF, SUMMARY OFF, COSTS OFF)
SELECT * FROM uuid_test WHERE id = '000e1df0-8c30-11e9-9bb8-e03f4977f7b7';
                                  QUERY PLAN                                  
------------------------------------------------------------------------------
 Index Only Scan using uidx_uuid_test_id on uuid_test (actual rows=1 loops=1)
   Index Cond: (id = '000e1df0-8c30-11e9-9bb8-e03f4977f7b7'::uuid)
   Heap Fetches: 1
(3 rows)

